ext/opcache/jit/zend_jit_x86.dasc

```diff
@@ -10836,6 +10836,45 @@ static int zend_jit_strlen(dasm_State **Dst, const zend_op *opline, const zend_o
 	return 1;
 }
 
+static int zend_jit_count(dasm_State **Dst, const zend_op *opline, const zend_op_array *op_array, uint32_t op1_info)
+{
+	zend_jit_addr res_addr = RES_ADDR();
+
+	if (opline->op1_type == IS_CONST) {
+		zval *zv;
+		size_t count;
+
+		zv = RT_CONSTANT(opline, opline->op1);
+		ZEND_ASSERT(Z_TYPE_P(zv) == IS_ARRAY);
+		count = zend_array_count(Z_ARRVAL_P(zv));
+
+		|	SET_ZVAL_LVAL res_addr, count
+		|	SET_ZVAL_TYPE_INFO res_addr, IS_LONG
+	} else {
+		zend_jit_addr op1_addr = OP1_ADDR();
+
+		ZEND_ASSERT((op1_info & (MAY_BE_UNDEF|MAY_BE_ANY|MAY_BE_REF)) == MAY_BE_ARRAY);
+		// Note: Conservatively always using zend_array_count because arrays may contain entries with IS_UNDEF at runtime, and op1_info doesn't represent that.
+		// Special cases for arrays include $GLOBALS, as well as arrays that were created from objects.
+		// (e.g. `$x = new self(); unset($x->declaredProp); echo count(get_object_vars($x))`).
+
+		|	GET_ZVAL_PTR FCARG1a, op1_addr
+		|	EXT_CALL zend_array_count, r0
+		|	SET_ZVAL_LVAL res_addr, r0
```

Comment:
This is wrong and should be `eax` (return value) for clarity - `r0` is the temporary register that would get used to store the function's address if needed in x86, not the return value. I probably also need to load FCARG1a again before calling ZVAL_PTR_DTOR in case it gets overridden? `|	GET_ZVAL_PTR FCARG1a, op1_addr`

Even after fixing that, this still gives the wrong address even for IS_CV.

Reply:
Oh. zend_array_count isn't ZEND_FASTCALL, but I'm passing it in the register that would be used for ZEND_FASTCALL. Other functions, such as zend_is_true, are FASTCALL.

`ZEND_API uint32_t zend_array_count(HashTable *ht)`
`ZEND_API int ZEND_FASTCALL zend_is_true(zval *op) /* {{{ */`

Reply:
It's okay to just mark it fastcall.

Reply:
> It's okay to just mark it fastcall.

I was wondering if it was deliberately left out, but the function declaration was added in 2015 in `6313e16a0472d42a2056364b1aacd7798d500b75` and wasn't changed since then.

After running `make clean` and rebuilding, the test passes on 32-bit platforms

Reply:
> This is wrong and should be eax (return value) for clarity - r0 is the temporary register that would get used to store the function's address if needed in x86,

https://corsix.github.io/dynasm-doc/instructions.html#registers

It turns out I had it right the first time - `r0` is actually the return value of `zend_array_count` (`rax` on 64-bit platforms, `eax` on 32-bit platforms)